docs/reference/configuration.txt
.. _configuration:

*************
Configuration
*************

.. default-domain:: mongodb

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

ActiveDocument is customarily configured through a ``active_document.yml`` file that specifies
options and clients. The simplest configuration is as follows, which configures
ActiveDocument to talk to a MongoDB server at "localhost:27017" and use the database
named "active_document".

.. code-block:: yaml

  development:
    clients:
      default:
        database: active_document
        hosts:
          - localhost:27017

The top level key in the configuration file, ``development`` in the above
example, refers to the environment name which the application is executing in,
i.e. ``development``, ``test`` or ``production``. The third level key,
``default`` in the above example, refers to the Mongo client name.
Most applications will use a single client named ``default``.


Generating Default Configuration
================================

If you are using Ruby on Rails, you can have ActiveDocument generate a default
configuration file for you by running the following command:

.. code-block:: bash

  rails g active_document:config

The configuration file will be placed in ``config/active_document.yml``. An
initializer will also be created and placed in
``config/initializers/active_document.rb``. It is recommended that all configuration
be specified in ``config/active_document.yml``, but if you prefer, the ``active_document.rb``
initializer may also be used to set configuration options. Note, though, that
settings in ``active_document.yml`` always take precedence over settings in the
initializer.

If you are not using Ruby on Rails, you can copy the minimal configuration
given above and save it as ``config/active_document.yml``.


Loading ActiveDocument Configuration
=============================

If you are using Ruby on Rails, ActiveDocument configuration is automatically loaded
for the current environment as stored in ``Rails.env`` when the application
loads.

You may need to configure the ORM for your application to be ActiveDocument by
adding the following to ``application.rb``:

.. code-block:: ruby

  config.generators do |g|
    g.orm :active_document
  end

If you are not using Ruby on Rails, ActiveDocument configuration must be loaded
manually. This can be done via the ``ActiveDocument.load!`` method, which takes
the configuration file path as its argument, as follows:

.. code-block:: ruby

  # Use automatically detected environment name
  ActiveDocument.load!('path/to/your/active_document.yml')

  # Specify environment name manually
  ActiveDocument.load!('path/to/your/active_document.yml', :production)

When ActiveDocument is asked to automatically detect the environment name,
it does so by examining the following sources, in order:

- If ``Rails`` top level constant is defined, ``Rails.env``.
- If ``Sinatra`` top level constant is defined, ``Sinatra::Base.environment``.
- The ``RACK_ENV`` environment variable.
- The ``MONGOID_ENV`` environment variable.

It is also possible to configure ActiveDocument directly in Ruby, without using
a configuration file. This configuration style does not support the concept
of environments - whatever configuration is provided, it is applied to the
current environment - but it does support defining multiple clients.

.. code-block:: ruby

  ActiveDocument.configure do |config|
    config.clients.default = {
      hosts: ['localhost:27017'],
      database: 'my_db',
    }

    config.log_level = :warn
  end

.. note::

  ActiveDocument must be configured *before* any component of it is used or referenced.
  Once a component is used or referenced, changing configuration may not apply
  changes to already instantiated components.


.. _configuration-options:

ActiveDocument Configuration Options
=============================

The following annotated example ``active_document.yml`` demonstrates how ActiveDocument
can be configured.

ActiveDocument delegates to the Ruby driver for client configuration. Please review
`the driver documentation <https://mongodb.com/docs/ruby-driver/current/reference/create-client/>`_
for details on driver options.

.. code-block:: yaml

  development:
    # Configure available database clients. (required)
    clients:
      # Defines the default client. (required)
      default:
        # ActiveDocument can connect to a URI accepted by the driver:
        # uri: mongodb://user:password@mongodb.domain.com:27017/my_db_development

        # Otherwise define the parameters separately.
        # This defines the name of the default database that ActiveDocument can connect to.
        # (required).
        database: my_db_development
        # Provides the hosts the default client can connect to. Must be an array
        # of host:port pairs. (required)
        hosts:
          - localhost:27017
        options:
          # Note that all options listed below are Ruby driver client options (the mongo gem).
          # Please refer to the driver documentation of the version of the mongo gem you are using
          # for the most up-to-date list of options.

          # Change the default write concern. (default = { w: 1 })
          # write:
          #   w: 1

          # Change the default read preference. Valid options for mode are: :secondary,
          # :secondary_preferred, :primary, :primary_preferred, :nearest
          # (default: primary)
          # read:
          #   mode: :secondary_preferred
          #   tag_sets:
          #     - use: web

          # The name of the user for authentication.
          # user: 'user'

          # The password of the user for authentication.
          # password: 'password'

          # The user's database roles.
          # roles:
          #   - 'dbOwner'

          # Change the default authentication mechanism. Valid options include:
          # :scram, :scram256, :mongodb_cr, :mongodb_x509, :gssapi, :aws, :plain.
          # MongoDB Server defaults to :scram, which will use "SCRAM-SHA-256" if available,
          # otherwise fallback to "SCRAM-SHA-1" (:scram256 will always use "SCRAM-SHA-256".)
          # This setting is handled by the MongoDB Ruby Driver. Please refer to:
          # https://mongodb.com/docs/ruby-driver/current/reference/authentication/
          # auth_mech: :scram

          # The database or source to authenticate the user against.
          # (default: the database specified above or admin)
          # auth_source: admin

          # Force a the driver cluster to behave in a certain manner instead of auto-
          # discovering. Can be one of: :direct, :replica_set, :sharded. Set to :direct
          # when connecting to hidden members of a replica set.
          # connect: :direct

          # Changes the default time in seconds the server monitors refresh their status
          # via hello commands. (default: 10)
          # heartbeat_frequency: 10

          # The time in seconds for selecting servers for a near read preference. (default: 0.015)
          # local_threshold: 0.015

          # The timeout in seconds for selecting a server for an operation. (default: 30)
          # server_selection_timeout: 30

          # The maximum number of connections in the connection pool. (default: 5)
          # max_pool_size: 5

          # The minimum number of connections in the connection pool. (default: 1)
          # min_pool_size: 1

          # The time to wait, in seconds, in the connection pool for a connection
          # to be checked in before timing out. (default: 5)
          # wait_queue_timeout: 5

          # The time to wait to establish a connection before timing out, in seconds.
          # (default: 10)
          # connect_timeout: 10

          # How long to wait for a response for each operation sent to the
          # server. This timeout should be set to a value larger than the
          # processing time for the longest operation that will be executed
          # by the application. Note that this is a client-side timeout;
          # the server may continue executing an operation after the client
          # aborts it with the SocketTimeout exception.
          # (default: nil, meaning no timeout)
          # socket_timeout: 5

          # The name of the replica set to connect to. Servers provided as seeds that do
          # not belong to this replica set will be ignored.
          # replica_set: name

          # Compressors to use for wire protocol compression. (default is to not use compression)
          # "zstd" requires zstd-ruby gem. "snappy" requires snappy gem.
          # Refer to: https://www.mongodb.com/docs/ruby-driver/current/reference/create-client/#compression
          # compressors: ["zstd", "snappy", "zlib"]

          # Whether to connect to the servers via ssl. (default: false)
          # ssl: true

          # The certificate file used to identify the connection against MongoDB.
          # ssl_cert: /path/to/my.cert

          # The private keyfile used to identify the connection against MongoDB.
          # Note that even if the key is stored in the same file as the certificate,
          # both need to be explicitly specified.
          # ssl_key: /path/to/my.key

          # A passphrase for the private key.
          # ssl_key_pass_phrase: password

          # Whether to do peer certification validation. (default: true)
          # ssl_verify: true

          # The file containing concatenated certificate authority certificates
          # used to validate certs passed from the other end of the connection.
          # ssl_ca_cert: /path/to/ca.cert

          # Whether to truncate long log lines. (default: true)
          # truncate_logs: true

    # Configure ActiveDocument-specific options. (optional)
    options:
      # Allow BSON::Decimal128 to be parsed and returned directly in
      # field values. When BSON 5 is present and the this option is set to false
      # (the default), BSON::Decimal128 values in the database will be returned
      # as BigDecimal.
      #
      # @note this option only has effect when BSON 5+ is present. Otherwise,
      #   the setting is ignored.
      # allow_bson5_decimal128: false

      # Application name that is printed to the MongoDB logs upon establishing
      # a connection. Note that the name cannot exceed 128 bytes in length.
      # It is also used as the database name if the database name is not
      # explicitly defined. (default: nil)
      # app_name: nil

      # When this flag is false, callbacks for embedded documents will not be
      # called. This is the default in 9.0.
      #
      # Setting this flag to true restores the pre-9.0 behavior, where callbacks
      # for embedded documents are called. This may lead to stack overflow errors
      # if there are more than around 1000 embedded documents in the root
      # document's dependencies graph.
      # See https://jira.mongodb.org/browse/MONGOID-5658 for more details.
      # around_callbacks_for_embeds: false

      # Sets the async_query_executor for the application. By default the thread pool executor
      # is set to `:immediate`. Options are:
      #
      #   - :immediate - Initializes a single +Concurrent::ImmediateExecutor+
      #   - :global_thread_pool - Initializes a single +Concurrent::ThreadPoolExecutor+
      #      that uses the +async_query_concurrency+ for the +max_threads+ value.
      # async_query_executor: :immediate

      # Mark belongs_to associations as required by default, so that saving a
      # model with a missing belongs_to association will trigger a validation
      # error.
      # belongs_to_required_by_default: true

      # Set the global discriminator key. (default: '_type')
      discriminator_key: '_type'

      # Raise an exception when a field is redefined.
      # duplicate_fields_exception: false

      # Defines how many asynchronous queries can be executed concurrently.
      # This option should be set only if `async_query_executor` is set
      # to `:global_thread_pool`.
      # global_executor_concurrency: nil

      # When this flag is true, any attempt to change the _id of a persisted
      # document will raise an exception (`Errors::ImmutableAttribute`).
      # This is the default in 9.0. Setting this flag to false restores the
      # pre-9.0 behavior, where changing the _id of a persisted
      # document might be ignored, or it might work, depending on the situation.
      # immutable_ids: true

      # Include the root model name in json serialization.
      # include_root_in_json: false

      # # Include the _type field in serialization.
      # include_type_for_serialization: false

      # Whether to join nested persistence contexts for atomic operations
      # to parent contexts by default.
      # join_contexts: false

      # When this flag is false (the default as of ActiveDocument 9.0), a document that
      # is created or loaded will remember the storage options that were active
      # when it was loaded, and will use those same options by default when
      # saving or reloading itself.
      #
      # When this flag is true you'll get pre-9.0 behavior, where a document will
      # not remember the storage options from when it was loaded/created, and
      # subsequent updates will need to explicitly set up those options each time.
      #
      # For example:
      #
      #    record = Model.with(collection: 'other_collection') { Model.first }
      #
      # This will try to load the first document from 'other_collection' and
      # instantiate it as a Model instance. Pre-9.0, the record object would
      # not remember that it came from 'other_collection', and attempts to
      # update it or reload it would fail unless you first remembered to
      # explicitly specify the collection every time.
      #
      # As of ActiveDocument 9.0, the record will remember that it came from
      # 'other_collection', and updates and reloads will automatically default
      # to that collection, for that record object.
      # legacy_persistence_context_behavior: false

      # When this flag is false, a document will become read-only only once the
      # #readonly! method is called, and an error will be raised on attempting
      # to save or update such documents, instead of just on delete. When this
      # flag is true, a document is only read-only if it has been projected
      # using #only or #without, and read-only documents will not be
      # deletable/destroyable, but they will be savable/updatable.
      # When this feature flag is turned on, the read-only state will be reset on
      # reload, but when it is turned off, it won't be.
      # legacy_readonly: false

      # The log level.
      #

docs/release-notes.txt
.. _release-notes:

*************
Release Notes
*************

.. default-domain:: mongodb

.. toctree::
  :titlesonly:

  release-notes/upgrading
  release-notes/active_document-9.0
  release-notes/active_document-8.1
  release-notes/active_document-8.0
  release-notes/active_document-7.5
  release-notes/active_document-7.4
  release-notes/active_document-7.3
  release-notes/active_document-7.2
  release-notes/active_document-7.1
  release-notes/active_document-7.0
  release-notes/active_document-6
  release-notes/active_document-5

Overview
--------

See the following sections to learn more about upgrading ActiveDocument:

- :ref:`Upgrading ActiveDocument <upgrading>`
- :ref:`ActiveDocument 9.0 <active_document-9.0>`
- :ref:`ActiveDocument 8.1 <active_document-8.1>`
- :ref:`ActiveDocument 8.0 <active_document-8.0>`
- :ref:`ActiveDocument 7.5 <active_document-7.5>`
- :ref:`ActiveDocument 7.4 <active_document-7.4>`
- :ref:`ActiveDocument 7.3 <active_document-7.3>`


lib/active_document/config.rb
# frozen_string_literal: true

require 'active_document/config/defaults'
require 'active_document/config/environment'
require 'active_document/config/options'
require 'active_document/config/validators'
require 'active_document/config/introspection'
require 'active_document/config/encryption'

module ActiveDocument

  # This module defines all the configuration options for ActiveDocument, including
  # the database connections.
  module Config
    extend Forwardable
    extend Options
    extend Defaults
    extend Encryption
    extend self

    def_delegators ::ActiveDocument, :logger, :logger=

    LOCK = Mutex.new

    # Application name that is printed to the MongoDB logs upon establishing
    # a connection. Note that the name cannot exceed 128 bytes in length.
    # It is also used as the database name if the database name is not
    # explicitly defined.
    option :app_name, default: nil

    # Mark belongs_to associations as required by default, so that saving a
    # model with a missing belongs_to association will trigger a validation
    # error.
    option :belongs_to_required_by_default, default: true

    # Set the global discriminator key.
    option :discriminator_key, default: '_type'

    # Raise an exception when a field is redefined.
    option :duplicate_fields_exception, default: false

    # Include the root model name in json serialization.
    option :include_root_in_json, default: false

    # # Include the _type field in serialization.
    option :include_type_for_serialization, default: false

    # Whether to join nested persistence contexts for atomic operations
    # to parent contexts by default.
    option :join_contexts, default: false

    # The log level.
    #
    # It must be set prior to referencing clients or Mongo.logger,
    # changes to this option are not be propagated to any clients and
    # loggers that already exist.
    #
    # Additionally, only when the clients are configured via the
    # configuration file is the log level given by this option honored.
    option :log_level, default: :info

    # Preload all models in development, needed when models use inheritance.
    option :preload_models, default: false

    # Raise an error when performing a #find and the document is not found.
    option :raise_not_found_error, default: true

    # Raise an error when defining a scope with the same name as an
    # existing method.
    option :scope_overwrite_exception, default: false

    # Store BigDecimals as Decimal128s instead of strings in the db.
    option :map_big_decimal_to_decimal128, default: true

    # Allow BSON::Decimal128 to be parsed and returned directly in
    # field values. When BSON 5 is present and this option is set to false
    # (the default), BSON::Decimal128 values in the database will be returned
    # as BigDecimal.
    #
    # @note this option only has effect when BSON 5+ is present. Otherwise,
    #   the setting is ignored.
    option :allow_bson5_decimal128, default: false, on_change: lambda { |allow|
      if BSON::VERSION >= '5.0.0'
        BSON::Registry.register(BSON::Decimal128::BSON_TYPE, allow ? BSON::Decimal128 : BigDecimal)
      end
    }

    # Sets the async_query_executor for the application. By default, the thread pool executor
    #   is set to `:immediate. Options are:
    #
    #   - :immediate - Initializes a single +Concurrent::ImmediateExecutor+
    #   - :global_thread_pool - Initializes a single +Concurrent::ThreadPoolExecutor+
    #      that uses the +global_executor_concurrency+ for the +max_threads+ value.
    option :async_query_executor, default: :immediate

    # Defines how many asynchronous queries can be executed concurrently.
    # This option should be set only if `async_query_executor` is set
    # to `:global_thread_pool`.
    option :global_executor_concurrency, default: nil

    VALID_ISOLATION_LEVELS = %i[rails thread fiber].freeze

    # Defines the isolation level that ActiveDocument uses to store its internal
    # state.
    #
    # Valid values are:
    # - `:rails` - Uses the isolation level that Rails currently has
    #  configured. (This is the default.)
    # - `:thread` - Uses thread-local storage.
    # - `:fiber` - Uses fiber-local storage (only supported in Ruby 3.2+).
    #
    # If set to `:fiber`, ActiveDocument will use fiber-local storage instead. This
    # may be necessary if you are using libraries like Falcon, which use
    # fibers to manage concurrency.
    #
    # Note that the `:fiber` isolation level is only supported in Ruby 3.2
    # and later, due to semantic differences in how fiber storage is handled
    # in earlier Ruby versions.
    option :isolation_level, default: :rails, on_change: lambda { |level|
      validate_isolation_level!(level)
    }

    # Returns the (potentially-dereferenced) isolation level that ActiveDocument
    # will use to store its internal state. If `isolation_level` is set to
    # `:rails`, this will return the isolation level that Rails is current
    # configured to use (`ActiveSupport::IsolatedExecutionState.isolation_level`).
    #
    # If using an older version of Rails that does not support
    # ActiveSupport::IsolatedExecutionState, this will return `:thread`
    # instead.
    #
    # @api private
    def real_isolation_level
      return isolation_level unless isolation_level == :rails

      ActiveSupport::IsolatedExecutionState.isolation_level.tap do |level|
        # We can't guarantee that Rails will always support the same
        # isolation levels as ActiveDocument, so we check here to make sure
        # it's something we can work with.
        validate_isolation_level!(level)
      end
    end

    # Checks to see if the provided isolation level is something that ActiveDocument
    # supports. Raises Errors::UnsupportedIsolationLevel if it is not.
    #
    # This will also raise an error if the isolation level is set to `:fiber`
    # and the Ruby version is less than 3.2, since fiber-local storage
    # is not supported in earlier Ruby versions.
    #
    # @api private
    def validate_isolation_level!(level)
      return if VALID_ISOLATION_LEVELS.include?(level)

      raise Errors::UnsupportedIsolationLevel.new(level)
    end

    # When this flag is false, a document will become read-only only once the
    # #readonly! method is called, and an error will be raised on attempting
    # to save or update such documents, instead of just on delete. When this
    # flag is true, a document is only read-only if it has been projected
    # using #only or #without, and read-only documents will not be
    # deletable/destroyable, but they will be savable/updatable.
    # When this feature flag is turned on, the read-only state will be reset on
    # reload, but when it is turned off, it won't be.
    option :legacy_readonly, default: false

    # When this flag is false (the default as of ActiveDocument 9.0), a document that
    # is created or loaded will remember the storage options that were active
    # when it was loaded, and will use those same options by default when
    # saving or reloading itself.
    #
    # When this flag is true you'll get pre-9.0 behavior, where a document will
    # not remember the storage options from when it was loaded/created, and
    # subsequent updates will need to explicitly set up those options each time.
    #
    # For example:
    #
    #    record = Model.with(collection: 'other_collection') { Model.first }
    #
    # This will try to load the first document from 'other_collection' and
    # instantiate it as a Model instance. Pre-9.0, the record object would
    # not remember that it came from 'other_collection', and attempts to
    # update it or reload it would fail unless you first remembered to
    # explicitly specify the collection every time.
    #
    # As of ActiveDocument 9.0, the record will remember that it came from
    # 'other_collection', and updates and reloads will automatically default
    # to that collection, for that record object.
    option :legacy_persistence_context_behavior, default: false

    # When this flag is true, any attempt to change the _id of a persisted
    # document will raise an exception (`Errors::ImmutableAttribute`).
    # This is the default in 9.0. Setting this flag to false restores the
    # pre-9.0 behavior, where changing the _id of a persisted
    # document might be ignored, or it might work, depending on the situation.
    option :immutable_ids, default: true

    # When this flag is true, callbacks for every embedded document will be
    # called only once, even if the embedded document is embedded in multiple
    # documents in the root document's dependencies graph.
    # This is the default in 9.0. Setting this flag to false restores the
    # pre-9.0 behavior, where callbacks are called for every occurrence of an
    # embedded document. The pre-9.0 behavior leads to a problem that for multi
    # level nested documents callbacks are called multiple times.
    # See https://jira.mongodb.org/browse/MONGOID-5542
    option :prevent_multiple_calls_of_embedded_callbacks, default: true

    # When this flag is false, callbacks for embedded documents will not be
    # called. This is the default in 9.0.
    #
    # Setting this flag to true restores the pre-9.0 behavior, where callbacks
    # for embedded documents are called. This may lead to stack overflow errors
    # if there are more than around 1000 embedded documents in the root
    # document's dependencies graph.
    # See https://jira.mongodb.org/browse/MONGOID-5658 for more details.
    option :around_callbacks_for_embeds, default: false

    # When this flag is false, named scopes cannot unset a default scope.
    # This is the traditional (and default) behavior in ActiveDocument 9 and earlier.
    #
    # Setting this flag to true will allow named scopes to unset the default
    # scope. This will be the default in ActiveDocument 10.
    #
    # See https://jira.mongodb.org/browse/MONGOID-5785 for more details.
    option :allow_scopes_to_unset_default_scope, default: false

    # When this flag is false, indexes are (roughly) validated on the client
    # to prevent duplicate indexes being declared. This validation is
    # incomplete, however, and can result in some indexes being silently
    # ignored.
    #
    # Setting this to true will allow duplicate indexes to be declared and sent
    # to the server. The server will then validate the indexes and raise an
    # exception if duplicates are detected.
    #
    # TODO: ActiveDocument: Actually implement the match logic properly.
    #
    # See https://jira.mongodb.org/browse/MONGOID-5827 for an example of the
    # consequences of duplicate index checking.
    option :allow_duplicate_index_declarations, default: false

    # When this flag is false (the default), modifying associations from the
    # inverse (has_*) side will raise an InverseRelationAssignmentDisallowed
    # error. For example, with `belongs_to_many :parks` on Dog and
    # `has_many :dogs` on Park, you must use `dog.parks << park` rather than
    # `park.dogs << dog`.
    #
    # Setting this flag to true allows assignment from the inverse side.
    # The FK change is queued and persisted when the base document is saved.
    #
    # This can also be enabled per-instance with `doc.allow_inverse_relation_assignment!`
    option :allow_inverse_relation_assignment, default: false

    # When this flag is true, performing a serializable_hash(only: []) will
    # result in all fields being included. This is the traditional (and default)
    # behavior in ActiveDocument 9 and earlier.
    #
    # Setting this flag to false will change the behavior to suppress all fields
    # in the serialized hash. This will be the default in ActiveDocument 10.
    #
    # See https://jira.mongodb.org/browse/MONGOID-5892 for more details.
    option :serializable_hash_with_legacy_only, default: true

    # Returns the Config singleton, for use in the configure DSL.
    #
    # @return [ self ] The Config singleton.
    def config
      self
    end

    # Has ActiveDocument been configured? This is checking that at least a valid
    # client config exists.
    #
    # @example Is ActiveDocument configured?
    #   config.configured?
    #
    # @return [ true | false ] If ActiveDocument is configured.
    def configured?
      clients.key?(:default)
    end

    # Connect to the provided database name on the default client.
    #
    # @note Use only in development or test environments for convenience.
    #
    # @example Set the database to connect to.
    #   config.connect_to("active_document_test")
    #
    # @param [ String ] name The database name.
    def connect_to(name, options = { read: { mode: :primary } })
      self.clients = {
        default: {
          database: name,
          hosts: ['localhost:27017'],
          options: options
        }
      }
    end

    # Return field names that could cause destructive things to happen if
    # defined in a ActiveDocument::Document.
    #
    # @example Get the destructive fields.
    #   config.destructive_fields
    #
    # @return [ Array<String> ] An array of bad field names.
    def destructive_fields
      Composable.prohibited_methods
    end

    # Load the settings from a compliant active_document.yml file. This can be used for
    # easy setup with frameworks other than Rails.
    #
    # @example Configure ActiveDocument.
    #   ActiveDocument.load!("/path/to/active_document.yml")
    #
    # @param [ String ] path The path to the file.
    # @param [ String | Symbol ] environment The environment to load.
    def load!(path, environment = nil)
      settings = Environment.load_yaml(path, environment)
      if settings.present?
        Clients.disconnect
        Clients.clear
        load_configuration(settings)
      end
      settings
    end

    # Get all the models in the application - this is everything that includes
    # ActiveDocument::Document.
    #
    # @example Get all the models.
    #   config.models
    #
    # @return [ Array<Class> ] All the models in the application.
    def models
      @models ||= []
    end

    # Register a model in the application with ActiveDocument.
    #
    # @example Register a model.
    #   config.register_model(Band)
    #
    # @param [ Class ] klass The model to register.
    def register_model(klass)
      LOCK.synchronize do
        models.push(klass) unless models.include?(klass)
      end
    end

    # Deregister a model in the application with ActiveDocument.
    #
    # @param [ Class ] klass The model to deregister.
    #
    # @api private
    def deregister_model(klass)
      LOCK.synchronize do
        models.delete(klass)
      end
    end

    # From a hash of settings, load all the configuration.
    #
    # @example Load the configuration.
    #   config.load_configuration(settings)
    #
    # @param [ Hash ] settings The configuration settings.
    def load_configuration(settings)
      configuration = settings.with_indifferent_access
      self.options = configuration[:options]
      self.clients = configuration[:clients]
      Mongo.options = configuration[:driver_options] || {}
      set_log_levels
    end

    # Override the database to use globally.
    #
    # @example Override the database globally.
    #   config.override_database(:optional)
    #
    # @param [ String | Symbol ] name The name of the database.
    #
    # @return [ String | Symbol ] The global override.
    def override_database(name)
      Threaded.database_override = name
    end

    # Override the client to use globally.
    #
    # @example Override the client globally.
    #   config.override_client(:optional)
    #
    # @param [ String | Symbol ] name The name of the client.
    #
    # @return [ String | Symbol ] The global override.
    def override_client(name)
      Threaded.client_override = name ? name.to_s : nil
    end

    # Purge all data in all collections, including indexes.
    #
    # @example Purge all data.
    #   ActiveDocument::Config.purge!
    #
    # @note This is the fastest way to drop all data.
    #
    # @return [ true ] true.
    def purge!
      global_client.database.collections.each(&:drop) and true
    end

    # Truncate all data in all collections, but not the indexes.
    #
    # @example Truncate all collection data.
    #   ActiveDocument::Config.truncate!
    #
    # @note This will be slower than purge!
    #
    # @return [ true ] true.
    def truncate!
      global_client.database.collections.each do |collection|
        collection.find.delete_many
      end and true
    end

    # Set the configuration options. Will validate each one individually.
    #
    # @example Set the options.
    #   config.options = { raise_not_found_error: true }
    #
    # @param [ Hash ] options The configuration options.
    def options=(options)
      return unless options

      Validators::AsyncQueryExecutor.validate(options)
      options.each_pair do |option, value|
        Validators::Option.validate(option)
        send(:"#{option}=", value)
      end
    end

    # Get the client configuration or an empty hash.
    #
    # @example Get the clients configuration.
    #   config.clients
    #
    # @return [ Hash ] The clients configuration.
    def clients
      @clients ||= {}
    end

    # Defines a field type mapping, for later use in field :type option.
    #
    # @example
    #   ActiveDocument.configure do |config|
    #     config.field_type :point, Point
    #   end
    #
    # @param [ Symbol | String ] type_name The identifier of the
    #   defined type. This identifier may be accessible as either a
    #   Symbol or a String regardless of the type passed to this method.
    # @param [ Module ] klass the class of the defined type, which must
    #   include mongoize, demongoize, and evolve methods.
    def field_type(type_name, klass)
      ActiveDocument::Fields::FieldTypes.define_type(type_name, klass)
    end

    # Defines an option for the field macro, which runs the handler
    # provided as a block.
    #
    # No assumptions are made about what functionality the handler might
    # perform, so it will always be called if the `option_name` key is
    # provided in the field definition -- even if it is false or nil.
    #
    # @example
    #   ActiveDocument.configure do |config|
    #     config.field_option :required do |model, field, value|
    #       model.validates_presence_of field.name if value
    #     end
    #   end
    #
    # @param [ Symbol ] option_name the option name to match against
    # @param [ Proc ] block the handler to execute when the option is
    #   provided.
    def field_option(option_name, &block)
      ActiveDocument::Fields.option(option_name, &block)
    end

    private

    def set_log_levels
      ActiveDocument.logger.level = ActiveDocument::Config.log_level unless defined?(::Rails)
      Mongo::Logger.logger.level = ActiveDocument.logger.level
    end

    def clients=(clients)
      raise Errors::NoClientsConfig unless clients

      c = clients.with_indifferent_access
      Validators::Client.validate(c)
      @clients = c
    end

    # Get database client that respects global overrides
    # Config.override_database and Config.override_client.
    #
    # @return [Mongo::Client] Client according to global overrides.
    def global_client
      client = if Threaded.client_override
                 Clients.with_name(Threaded.client_override)
               else
                 Clients.default
               end
      if Threaded.database_override
        client.use(Threaded.database_override)
      else
        client
      end
    end

    # Mixin used to deprecate ActiveDocument configuration options.
    module DeprecatedOptions
      OPTIONS = %i[].freeze

      OPTIONS.each do |option|
        define_method(:"#{option}=") do |value|
          ActiveDocument::Warnings.send(:"warn_#{option}_deprecated")
          super(value)
        end
      end
    end

    prepend DeprecatedOptions
  end
end
